### PR TITLE
Remove cast of y_true to y_pred data type in sparse categorical cross entropy loss

### PR DESCRIPTION
**Pull request by old-school-kid**
Casting y_true to y_pred data type gives erroneous loss. Like in case of Mean Squared error if y_true is > 2**32 and y_pred data type is int32 then resulting loss will be wrong.
A detailed [notebook](https://colab.research.google.com/drive/1xdWCu4MavfmfimQQlXciPl_Atbip8fWi?usp=sharing) on the error which also happens in mean absolute loss,  categorical cross entropy loss and sparse categorical cross entropy loss and also contains the proposed changes. 

The issue was raised [here](https://github.com/keras-team/keras/issues/15014#issue-954672651)

### CODE REVIEW

**Review by qlzh727 — CHANGES_REQUESTED** [review]
I think the cast is mainly trying to address the dtype mismatch, eg int and float for label and prediction, which is quite common in normal case. The 2*32 seems to be a very extreme case, which we barely hit. Can we change the PR to only cast the dtype if the label and prediction are in different category?

**Review by qlzh727 — CHANGES_REQUESTED** [review]
I think we should only remove the cast for sparse_categorical_crossentropy(), since the label value could be large based on the dimension of the prediction. The rest of them like binary_crossentropy or categorical_crossentropy, the label value is either one_hot or just 0 and 1, which won't be affected when casting.

Also since backend.sparse_categorical_crossentropy will cast the y_true to int64 anyway, removing the y_true cast here is correct.
https://github.com/keras-team/keras/blob/00518dcf7f5b7de7064267979d7e9f9916e777eb/keras/backend.py#L4952

File: keras/losses.py [hunk 1]
```diff
@@ -1536,7 +1536,7 @@ def categorical_hinge(y_true, y_pred):
     Categorical hinge loss values.
   """
   y_pred = tf.convert_to_tensor(y_pred)
-  y_true = tf.cast(y_true, y_pred.dtype)
```
Comment (qlzh727) [inline keras/losses.py:1539]: I think you should keep this cast here, since the y_true are expect to be either `{-1, +1}` or `{0, 1}` (i.e. a one-hot-encoded tensor). See the docstring.

We probably want to fix the example in the docstring since the y_true is given in the range of 0-3.

Comment (old-school-kid) [reply keras/losses.py]: Yeah made the changes as requested.

**Review by qlzh727 — APPROVED** [review]
Thanks for the fix.
